lib/addons/argocd/index.ts

```diff
@@ -1,26 +1,359 @@
+    adminPasswordSecretName?: string,
 
+    /**
+     * Values to pass to the chart as per https://github.com/argoproj/argo-helm/blob/master/charts/argo-cd/values.yaml.
+     */
+    values?: {
+        [key: string]: any;
+    };
+}
+
+
+const argoDefaults: ArgoCDAddOnProps = {
+    namespace: "argocd"
+}
+
+/**
+ * Local function to create a secret reference for SSH key.
+ * @param url 
+ * @param secretName 
+ * @returns 
+ */
+const sshRepoRef = (url: string, secretName: string): string => yaml.stringify(
+    [{
+        url,
+        sshPrivateKeySecret: {
+            name: secretName,
+            key: "sshPrivateKey"
+        }
+    }]
+);
+
+/**
+ * Local function to a secret reference for username/pwd or username/token key.
+ * @param url 
+ * @param secretName 
+ * @returns 
+ */
+const userNameRepoRef = (url: string, secretName: string): string => yaml.stringify(
+    [{
+        url,
+        usernameSecret: {
+            name: secretName,
+            key: "username"
+        },
+        passwordSecret: {
+            name: secretName,
+            key: "password"
+        }
+    }]
+);
+
+/**
+ * Implementation of ArgoCD add-on and post deployment hook.
+ */
 export class ArgoCDAddOn implements ClusterAddOn, ClusterPostDeploy {
 
-    readonly namespace: string;
+    readonly options: ArgoCDAddOnProps;
+    private chartNode: HelmChart;
 
-    constructor(namespace?: string) {
-        this.namespace = namespace ?? "argocd";
+    constructor(props?: ArgoCDAddOnProps) {
+        this.options = { ...argoDefaults, ...props };
     }
 
-    deploy(clusterInfo: ClusterInfo): void {
-        clusterInfo.cluster.addHelmChart("argocd-addon", {
+    /**
+     * Impementation of the add-on contract deploy method.
+    */
+    async deploy(clusterInfo: ClusterInfo): Promise<Construct> {
+
+        const namespace = this.createNamespace(clusterInfo);
+
+        const sa = this.createServiceAccount(clusterInfo);
+        sa.node.addDependency(namespace);
+
+        const repo = await this.createSecretKey(clusterInfo, namespace);
+
+        const values = this.options.values ?? {
+            server: {
+                serviceAccount: {
+                    create: false
+                },
+                config: {
+                    repositories: repo
+                }
+            }
+        };
+
+        if(this.options.adminPasswordSecretName) {
+            const adminSecret = await this.createAdminSecret(clusterInfo.cluster.stack.region);
+            values['configs'] = {
+                secret: {
+                    argocdServerAdminPassword: adminSecret
+                }
+            }
+        }
+
+        this.chartNode = clusterInfo.cluster.addHelmChart("argocd-addon", {
             chart: "argo-cd",
-            
+            release: "ssp-addon",
             repository: "https://argoproj.github.io/argo-helm",
-            version: '3.2.3',
-            namespace: this.namespace,
+            version: '3.10.0',
```

Comment:
This applies across the board. If we don't test with another version, chances are, our logic won't work because of argo changes. It also makes it less supportable. Hence, fixed version of the chart is mandatory across the board, if we shoot for a managed solution. Let's discuss if needed.

Reply:
That makes sense.